core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/chunkedencoding/ChunkedEncodedInputStream.java

```diff
@@ -0,0 +1,293 @@
+ * This implementation supports chunk-headers, chunk-extensions, and trailers.
+ */
+@SdkProtectedApi
+public final class ChunkedEncodedInputStream extends SdkInputStream {
+    private static final Logger LOG = Logger.loggerFor(ChunkedEncodedInputStream.class);
+    private static final byte[] CRLF = {'\r', '\n'};
+    private static final byte[] END = {};
+
+    private final InputStream inputStream;
+    private final int chunkSize;
+
+    private final ChunkHeader header;
+    private final List<ChunkExtension> extensions = new ArrayList<>();
+    private final List<Trailer> trailers = new ArrayList<>();
+
+    private Chunk currentChunk;
+    private boolean isFinished = false;
+
+    private ChunkedEncodedInputStream(BuilderImpl builder) {
+        this.inputStream = Validate.notNull(builder.inputStream, "Input-Stream cannot be null!");
+        this.chunkSize = Validate.isPositive(builder.chunkSize, "Chunk-size must be greater than 0!");
+        this.header = Validate.notNull(builder.header, "Header cannot be null!");
+        this.extensions.addAll(Validate.notNull(builder.extensions, "Extensions cannot be null!"));
+        this.trailers.addAll(Validate.notNull(builder.trailers, "Trailers cannot be null!"));
+    }
+
+    @Override
+    protected InputStream getWrappedInputStream() {
+        return inputStream;
+    }
+
+    @Override
+    public int read() throws IOException {
+        if (currentChunk == null || currentChunk.hasEnded() && !isFinished) {
+            currentChunk = getChunk(inputStream);
+        }
+
+        return currentChunk.stream().read();
+    }
+
+    /**
+     * Get an encoded chunk from the input-stream, or the final chunk if we've reached the end of the input-stream.
+     */
+    private Chunk getChunk(InputStream stream) throws IOException {
+        LOG.debug(() -> "Reading next chunk.");
+        if (currentChunk != null) {
+            currentChunk.close();
+        }
+        // we *have* to read from the backing stream in order to figure out if it's the end or not
+        byte[] chunkData = new byte[chunkSize];
+        int read = read(stream, chunkData, chunkSize);
+
+        if (read > 0) {
+            // set the current chunk to the newly written chunk
+            return getNextChunk(Arrays.copyOf(chunkData, read));
+        }
+
+        LOG.debug(() -> "End of backing stream reached. Reading final chunk.");
+        isFinished = true;
+        // set the current chunk to the written final chunk
+        return getFinalChunk();
+    }
+
+    /**
+     * Read from an input-stream, up to a max number of bytes, storing them in a byte-array.
+     * The actual number of bytes can be less than the max in the event that we reach the end of the stream.
+     * <p>
+     * This method is necessary because we cannot assume the backing stream uses the default implementation of
+     * {@code read(byte b[], int off, int len)}
```

Comment:
It's also less efficient if the source stream is array-backed because it could do a `System.arraycopy`, which is faster than copying byte-by-byte.

Can we do this instead?
```
        int read = 0;
        int offset = 0;
        while (read > 0 && offset < maxBytesToRead) {
            read = inputStream.read(buf, offset, maxBytesToRead - offset);
            assert read != 0;
            offset += read;
        }

        return offset;
``` 

`InputStream.read(byte b[], int off, int len)` is guaranteed to return non-zero (assuming `len` is positive), so it *should* be equivalent, but faster for common source stream types, like `ByteArrayInputStream`.